include/up-cpp/transport/datamodel/UPayload.h

```diff
@@ -0,0 +1,119 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * 
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2024 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _UPAYLOAD_H_
+#define _UPAYLOAD_H_
+
+#include <memory>
+#include <cstdint>
+#include <cstring>
+
+namespace uprotocol::utransport {
+
+    enum class UPayloadType {
+        /** data passed by value - will be copied */
+        VALUE = 0,
+        /** data passed by reference - the user need to ensure that the
+         *  reference is valid until data is sent */
+        REFERENCE,
+        /** data passed by shared pointer */
+        SHARED,
+        /** invalid */
+        UNDEFINED
+    };
+
+    // The Serialization format for the data stored in UPayload.
+    enum UPayloadFormat {
+        /** Payload format was not is not set */
+        UNSPECIFIED = 0,
+        /** Payload is an Any protobuf message that contains the packed
+         *  payload */
+        PROTOBUF_WRAPPED_IN_ANY = 1,
+        /** Protobuf serialization format */
+        PROTOBUF = 2,
+        /** JSON serialization format */
+        JSON = 3,
+        /** Basic SOME/IP serialization format */
+        SOMEIP = 4,
+        /** SOME/IP TLV format */
+        SOMEIP_TLV = 5,
+        /** RAW (binary) format */
+        RAW = 6,
+        /** Text format */
+        TEXT = 7
+    };          
+
+    /**
+    * The UPayload contains the clean Payload information at its raw serialized
+    * structure of a byte[]
+    */
+    class UPayload {
+
+        public:
+            // Constructor
+            UPayload(const uint8_t* ptr, 
+                     const size_t size, 
+                     const UPayloadType &type);
+
+            // Copy constructor
+            UPayload(const UPayload& other);
+
+            // Assignment operator
+            UPayload& operator=(const UPayload& other);
+
+            // Move constructor
+            UPayload(UPayload&& other) noexcept;
+
+            // Move assignment operator
+            UPayload& operator=(UPayload&& other) noexcept;
+
+            void setFormat(const UPayloadFormat &format);
+
+            /**
+            * @return data
+            */
+            const uint8_t* data() const;
+
+            /**
+            * @return size
+            */
+            size_t size() const;
+
+           /**
+            * @return format type
+            */
+            UPayloadFormat format() const;
+            
+            bool isEmpty() const;
+
+        private:
+            std::shared_ptr<const uint8_t[]> dataPtr_;
+            size_t dataSize_;
+            UPayloadType type_;
+            UPayloadFormat payloadFormat_ = UPayloadFormat::RAW;
```

Comment:
Per spec UPAYLOAD_FORMAT_PROTOBUF_STUFFED_IN_ANY is the default and assumed when UNSPECIFIED is set.

Reply:
I don't like the fact that we marking data type information using enums instead of C++ types. Modern C++ style favors the use of std::variant<>, std::holds_alternative<>, std::get<>, std::visit for dealing with type information. Protobuf does not use this, of course, and that is a strong argument for not exposing protobuf to library clients. (Protobuf optional should be encapsulated such that client code only uses std::optional<> as well.)

Reply:
@debruce - `UPayloadFormat` and _types_ aren't exactly equivalent. The payload format describes a data serialization format (e.g. json or protobuf) without explicitly declaring what type in C++ would be used to represent that. At the `UPayload` layer, we are talking about a sequence of bytes accompanied by a hint as to how one would decode those bytes.

I do agree that generally it is better in C++ for different _kinds_ of things to be represented as different _types_. I address that, to a certain extent, in the API updates I'm working on using the feedback here in this PR. In the [`UMessageBuilder`](https://github.com/gregmedd/up-cpp/blob/rework/review/api_WIP/include/up-cpp/datamodel/builder/UMessage.h#L247) (based heavily on the `up-rust` interface of the same name), there is an interface that would be used something like this:

```
void handleRpcRequest(const UMessage& request) {
    Json response_payload = makeResponsePayload(request);

    auto response =
            UMessageBuilder::respondTo(request)
                .buildWithPayloadValue<JsonSerializer>(response_payload);

    transport->send(response);
}
```

In this way, the type information is preserved up until the point of serialization. The `JsonSerializer` here is just an object with a static `serialize()` interface that turns a `Json` (which we can assume is some object representing JSON-compatible data) into a `std::vector<uint8_t>` and provides the correct `UPayloadFormat` to go with that operation.

As @PLeVasseur has pointed out, the `UMessageBuilder` in `up-rust` strikes a really nice balance. It doesn't hide the protobuf messages from the user completely, but its interfaces provide streamlined operations for the most common message types such that direct interaction with most fields in the protobuf messages are unnecessary.

----

Having said all that, if you have a proposal for an alternate interface, _now_ would be the time to share it. I'm getting pretty close to done rewriting the API and it'll get a lot harder to make big changes soon.

Do you have an example of the interface that would be used for `UMessage` or `UAttributes`?